File: go/binarysearch/74_search_matrix.go
package binarysearch

// searchMatrix 搜索二维矩阵
func searchMatrix(matrix [][]int, target int) bool {

	m, n := len(matrix), len(matrix[0])
	l, r := 0, m*n-1

	for l <= r {
		mid := l + (r-l)>>1

		// 对列求差 和 求余 分别得到行列坐标
		row, col := mid/n, mid%n

		if matrix[row][col] == target {
			return true
		} else if matrix[row][col] < target {
			l = mid + 1
		} else {
			r = mid - 1
		}

	}

	return false

}

File: rust/src/binarysearch/_74_search_matrix.rs
/// 搜索二维矩阵
pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
    let (m, n) = (matrix.len(), matrix[0].len());

    let (mut l, mut r) = (0, m * n - 1);

    while l <= r {
        let mid = l + ((r - l) >> 1);

        let (row, col) = (mid / n, mid % n);
        if matrix[row][col] == target {
            return true;
        } else if matrix[row][col] < target {
            l = mid + 1
        } else {
            // mid 是usize， 防止mid=0 的时候 mid -1 越界
            if mid == 0 {
                return false;
            }
            r = mid - 1
        }
    }

    false
}
